### PR TITLE
gh-107298: Fix numerous ref errors and typos in the C API docs

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-107298

----
:books: Documentation preview :books:: https://cpython-previews--108258.org.readthedocs.build/

### CODE REVIEW

File: Doc/c-api/exceptions.rst [hunk 1]
```diff
@@ -222,17 +222,21 @@ For convenience, some of these functions will always return a
...
 .. c:function:: PyObject* PyErr_SetFromWindowsErrWithFilename(int ierr, const char *filename)
 
-   Similar to :c:func:`PyErr_SetFromWindowsErrWithFilenameObject`, but the
-   filename is given as a C string.  *filename* is decoded from the filesystem
-   encoding (:func:`os.fsdecode`).
+   Similar to :c:func:`PyErr_SetFromWindowsErr`, with the additional behavior
+   that if *filename* is not ``NULL``, it is decoded from the filesystem
+   encoding (:func:`os.fsdecode`) and passed to the constructor of
+   :exc:`OSError` as a third parameter to be used to define the
+   :attr:`!filename` attribute of the exception instance.
```
Comment (vstinner) [inline Doc/c-api/exceptions.rst:229]: I'm not sure if Doc/c-api/exceptions.rst changes are related or not? They don't seem to fix Sphinx warnings.

**Review by vstinner — APPROVED** [review]
LGTM. But I'm not sure if Doc/c-api/exceptions.rst changes are related.

File: Doc/c-api/exceptions.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/c-api/exceptions.rst:229]: Both descriptions referred non-existing function
```c
PyObject* PyErr_SetFromWindowsErrWithFilenameObject(int ierr, PyObject *filenameObject);
```
There is no, and never was such function.

So I rewrote the descriptions by referring other functions, using descriptions of other functions which add the `filename` parameter as an example.

It fixes 2 of 10 warnings in `Doc/c-api/exceptions.rst`.

Comment (vstinner) [reply Doc/c-api/exceptions.rst:229]: Oh I didn't notice, ok. Well, I approved your PR ;-)

**Pull request merged**